File: client/client-multi/client-objective-c/KaaTests/channel/DefaultBootstrapChannelTest.m
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#define HC_SHORTHAND
#import <OCHamcrest/OCHamcrest.h>

#define MOCKITO_SHORTHAND
#import <OCMockito/OCMockito.h>

#import <XCTest/XCTest.h>
#import <Foundation/Foundation.h>
#import "DefaultBootstrapChannel.h"
#import "AbstractHttpChannel.h"
#import "KaaClientState.h"
#import "GenericTransportInfo.h"
#import "TransportProtocolIdHolder.h"
#import "KeyUtils.h"
#import "TestsHelper.h"
#import "TransportProtocolIdHolder.h"

@interface HttpClientMockWithEncoder : HttpClientMock

@end

@implementation HttpClientMockWithEncoder

- (MessageEncoderDecoder *)getEncoderDecoder {
    MessageEncoderDecoder *crypt = mock([MessageEncoderDecoder class]);
    @try {
        [given([crypt decodeData:anything()]) willReturn:[TestsHelper getData]];
    }
    @catch (NSException *exception) {
        NSLog(@"Error decoding data: %@, reason: %@", exception.name, exception.reason);
    }
    return crypt;
}

@end


@interface DefaultBootStrapChannelMock : DefaultBootstrapChannel

@property (nonatomic) NSInteger wantedNumberOfInvocations;

- (instancetype)initWithClient:(AbstractKaaClient *)client
                         state:(id<KaaClientState>)state
               failoverManager:(id<FailoverManager>)manager
     wantedNumberOfInvocations:(NSInteger)wantedNumberOfInvocations;

@end

@implementation DefaultBootStrapChannelMock

- (instancetype)initWithClient:(AbstractKaaClient *)client
                         state:(id<KaaClientState>)state
               failoverManager:(id<FailoverManager>)manager
     wantedNumberOfInvocations:(NSInteger)wantedNumberOfInvocations {
    self = [super initWithClient:client state:state failoverManager:manager];
    self.wantedNumberOfInvocations = wantedNumberOfInvocations;
    return self;
}

- (NSOperationQueue *)createExecutor {
    return [super createExecutor];
}

@end


@interface DefaultBootstrapChannelTest : XCTestCase

@end

@implementation DefaultBootstrapChannelTest

- (void)testChannelGetters {
    AbstractKaaClient *client = mock([AbstractKaaClient class]);
    id<KaaClientState> state = mockProtocol(@protocol(KaaClientState));
    id<FailoverManager> failoverManager = mockProtocol(@protocol(FailoverManager));
    id<KaaDataChannel> channel = [[DefaultBootstrapChannel alloc] initWithClient:client state:state failoverManager:failoverManager];
    
    NSDictionary *transportTypes = [channel getSupportedTransportTypes];
    
    XCTAssertEqualObjects(transportTypes[@(TRANSPORT_TYPE_BOOTSTRAP)], @(CHANNEL_DIRECTION_BIDIRECTIONAL));
    XCTAssertTrue([[TransportProtocolIdHolder HTTPTransportID] isEqual:[channel getTransportProtocolId]]);
    XCTAssertTrue([@"default_bootstrap_channel" isEqualToString:[channel getId]]);
}

- (void)testChannelSync {
    id<KaaChannelManager> manager = mockProtocol(@protocol(KaaChannelManager));
    AbstractHttpClient *httpClient = [[HttpClientMockWithEncoder alloc] init];
    id<FailoverManager> failoverManager = mockProtocol(@protocol(FailoverManager));
    
    [KeyUtils generateKeyPair];
    AbstractKaaClient *client = mock([AbstractKaaClient class]);
    [given([client createHttpClientWithURLString:anything() privateKeyRef:nil publicKeyRef:nil remoteKey:anything()]) willReturn:httpClient];
    [given([client getChannelManager]) willReturn:manager];
    
    id<KaaClientState> state = mockProtocol(@protocol(KaaClientState));
    id<KaaDataMultiplexer> multiplexer = mockProtocol(@protocol(KaaDataMultiplexer));
    id<KaaDataDemultiplexer> demultiplexer = mockProtocol(@protocol(KaaDataDemultiplexer));
    DefaultBootStrapChannelMock *channel = [[DefaultBootStrapChannelMock alloc] initWithClient:client state:state failoverManager:failoverManager wantedNumberOfInvocations:2];
    
    id<TransportConnectionInfo> server = [self createTestServerInfoWithServerType:SERVER_BOOTSTRAP transportProtocolId:[TransportProtocolIdHolder TCPTransportID] host:@"localhost" port:9889 publicKey:[KeyUtils getPublicKey]];
    [channel setServer:server];

    [channel syncForTransportType:TRANSPORT_TYPE_BOOTSTRAP];
    [channel setDemultiplexer:demultiplexer];
    [channel setDemultiplexer:nil];
    [channel syncForTransportType:TRANSPORT_TYPE_BOOTSTRAP];
    [channel setMultiplexer:multiplexer];
    [channel setMultiplexer:nil];
    [channel syncForTransportType:TRANSPORT_TYPE_CONFIGURATION];
    [channel syncForTransportType:TRANSPORT_TYPE_BOOTSTRAP];

    [NSThread sleepForTimeInterval:1];
    [verifyCount([channel getDemultiplexer], times(channel.wantedNumberOfInvocations)) processResponse:[TestsHelper getData]];
    [verifyCount([channel getMultiplexer], times(channel.wantedNumberOfInvocations)) compileRequestForTypes:anything()];
}

- (void)testShutDown {
    id<KaaChannelManager> manager = mockProtocol(@protocol(KaaChannelManager));
    AbstractHttpClient *httpClient = mock([AbstractHttpClient class]);
    id<FailoverManager> failoverManager = mockProtocol(@protocol(FailoverManager));
    
    AbstractKaaClient *client = mock([AbstractKaaClient class]);
    [given([client createHttpClientWithURLString:anything() privateKeyRef:[KeyUtils getPrivateKeyRef] publicKeyRef:[KeyUtils getPublicKeyRef] remoteKey:anything()]) willReturn:httpClient];
    [given([client getChannelManager]) willReturn:manager];
    
    id<KaaClientState> state = mockProtocol(@protocol(KaaClientState));
    id<KaaDataMultiplexer> multiplexer = mockProtocol(@protocol(KaaDataMultiplexer));
    id<KaaDataDemultiplexer> demultiplexer = mockProtocol(@protocol(KaaDataDemultiplexer));
    DefaultBootStrapChannelMock *channel = [[DefaultBootStrapChannelMock alloc] initWithClient:client state:state failoverManager:failoverManager wantedNumberOfInvocations:0];
    [channel setMultiplexer:multiplexer];
    [channel setDemultiplexer:demultiplexer];
    [channel shutdown];
    
    id<TransportConnectionInfo> server = [self createTestServerInfoWithServerType:SERVER_BOOTSTRAP transportProtocolId:[TransportProtocolIdHolder TCPTransportID] host:@"localhost" port:9889 publicKey:[KeyUtils getPublicKey]];
    [channel setServer:server];
    
    [channel syncForTransportType:TRANSPORT_TYPE_BOOTSTRAP];
    [channel syncAll];
    
    [NSThread sleepForTimeInterval:1];
    [verifyCount([channel getDemultiplexer], times(channel.wantedNumberOfInvocations)) processResponse:[TestsHelper getData]];
    [verifyCount([channel getMultiplexer], times(channel.wantedNumberOfInvocations)) compileRequestForTypes:anything()];
}

#pragma mark - Supporting methods

- (id<TransportConnectionInfo>) createTestServerInfoWithServerType:(ServerType)serverType transportProtocolId:(TransportProtocolId *)TPid host:(NSString *)host port:(uint32_t)port publicKey:(NSData *)publicKey {
    ProtocolMetaData *md = [TestsHelper buildMetaDataWithTransportProtocolId:TPid host:host port:port publicKey:publicKey];
    return  [[GenericTransportInfo alloc] initWithServerType:serverType meta:md];
}

@end

File: client/client-multi/client-java-core/src/test/java/org/kaaproject/kaa/client/channel/DefaultBootstrapChannelTest.java
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.kaaproject.kaa.client.channel;

import java.security.GeneralSecurityException;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.util.LinkedHashMap;
import java.util.concurrent.ExecutorService;

import org.junit.Assert;
import org.junit.Test;
import org.kaaproject.kaa.client.AbstractKaaClient;
import org.kaaproject.kaa.client.channel.failover.FailoverManager;
import org.kaaproject.kaa.client.channel.impl.channels.DefaultBootstrapChannel;
import org.kaaproject.kaa.client.persistence.KaaClientState;
import org.kaaproject.kaa.client.transport.AbstractHttpClient;
import org.kaaproject.kaa.common.TransportType;
import org.kaaproject.kaa.common.endpoint.security.KeyUtil;
import org.kaaproject.kaa.common.endpoint.security.MessageEncoderDecoder;
import org.mockito.Mockito;

public class DefaultBootstrapChannelTest {

    public ExecutorService fakeExecutor = new FakeExecutorService();

    class DefaultBootstrapChannelFake extends DefaultBootstrapChannel {

        private final int wantedNumberOfInvocations;

        public DefaultBootstrapChannelFake(AbstractKaaClient client, KaaClientState state,
                                           FailoverManager failoverManager, int wantedNumberOfInvocations) {
            super(client, state, failoverManager);
            this.wantedNumberOfInvocations = wantedNumberOfInvocations;
        }

        @Override
        protected ExecutorService createExecutor() {
            super.createExecutor();
            return fakeExecutor;
        }

        public void verify() throws Exception {
            Mockito.verify(getMultiplexer(), Mockito.times(wantedNumberOfInvocations)).compileRequest(Mockito.anyMap());
            Mockito.verify(getDemultiplexer(), Mockito.times(wantedNumberOfInvocations))
                    .processResponse(Mockito.eq(new byte[] { 5, 5, 5 }));
        }
        

    }
    
    class DefaultBootstrapChannelMock extends DefaultBootstrapChannelFake {
        
        public DefaultBootstrapChannelMock(AbstractKaaClient client, KaaClientState state,
                                           FailoverManager failoverManager, int wantedNumberOfInvocations) {
            super(client, state, failoverManager, wantedNumberOfInvocations);
        }
        
        protected AbstractHttpClient getHttpClient() {
            AbstractHttpClient client = Mockito.mock(AbstractHttpClient.class);
            MessageEncoderDecoder crypt = Mockito.mock(MessageEncoderDecoder.class);
            try {
                Mockito.when(crypt.decodeData(Mockito.any(byte[].class))).thenReturn(new byte[] { 5, 5, 5 });
            } catch (GeneralSecurityException e) {
                // TODO Auto-generated catch block
                e.printStackTrace();
            }
            Mockito.when(client.getEncoderDecoder()).thenReturn(crypt);
            
            return client;
        }
    }
    

    @Test
    public void testChannelGetters() {
        AbstractKaaClient client = Mockito.mock(AbstractKaaClient.class);
        KaaClientState state = Mockito.mock(KaaClientState.class);
        FailoverManager failoverManager = Mockito.mock(FailoverManager.class);
        KaaDataChannel channel = new DefaultBootstrapChannel(client, state, failoverManager);

        Assert.assertEquals(ChannelDirection.BIDIRECTIONAL, channel.getSupportedTransportTypes().get(TransportType.BOOTSTRAP));
        Assert.assertEquals(TransportProtocolIdConstants.HTTP_TRANSPORT_ID, channel.getTransportProtocolId());
        Assert.assertEquals("default_bootstrap_channel", channel.getId());
    }

    @Test
    public void testChannelSync() throws Exception {
        KaaChannelManager manager = Mockito.mock(KaaChannelManager.class);
        AbstractHttpClient httpClient = Mockito.mock(AbstractHttpClient.class);
        FailoverManager failoverManager = Mockito.mock(FailoverManager.class);
        Mockito.when(httpClient.executeHttpRequest(Mockito.anyString(), Mockito.any(LinkedHashMap.class), Mockito.anyBoolean()))
                .thenReturn(new byte[] { 5, 5, 5 });

        AbstractKaaClient client = Mockito.mock(AbstractKaaClient.class);
        Mockito.when(
                client.createHttpClient(Mockito.anyString(), Mockito.any(PrivateKey.class), Mockito.any(PublicKey.class),
                        Mockito.any(PublicKey.class))).thenReturn(httpClient);

        Mockito.when(client.getChannelManager()).thenReturn(manager);

        KaaClientState state = Mockito.mock(KaaClientState.class);
        KaaDataMultiplexer multiplexer = Mockito.mock(KaaDataMultiplexer.class);
        KaaDataDemultiplexer demultiplexer = Mockito.mock(KaaDataDemultiplexer.class);
        DefaultBootstrapChannelMock channel = new DefaultBootstrapChannelMock(client, state, failoverManager, 2);

        TransportConnectionInfo server = IPTransportInfoTest.createTestServerInfo(ServerType.BOOTSTRAP, TransportProtocolIdConstants.HTTP_TRANSPORT_ID,
                "localhost", 9889, KeyUtil.generateKeyPair().getPublic());

        channel.setServer(server);

        channel.sync(TransportType.BOOTSTRAP);
        channel.setDemultiplexer(demultiplexer);
        channel.setDemultiplexer(null);
        channel.sync(TransportType.BOOTSTRAP);
        channel.setMultiplexer(multiplexer);
        channel.setMultiplexer(null);
        channel.sync(TransportType.CONFIGURATION);
        channel.sync(TransportType.BOOTSTRAP);

        channel.verify();
    }


    @Test
    public void testShutdown() throws Exception {
        KaaChannelManager manager = Mockito.mock(KaaChannelManager.class);
        AbstractHttpClient httpClient = Mockito.mock(AbstractHttpClient.class);
        FailoverManager failoverManager = Mockito.mock(FailoverManager.class);
        Mockito.when(httpClient.executeHttpRequest(Mockito.anyString(), Mockito.any(LinkedHashMap.class), Mockito.anyBoolean())).thenThrow(
                new Exception());

        AbstractKaaClient client = Mockito.mock(AbstractKaaClient.class);
        Mockito.when(
                client.createHttpClient(Mockito.anyString(), Mockito.any(PrivateKey.class), Mockito.any(PublicKey.class),
                        Mockito.any(PublicKey.class))).thenReturn(httpClient);

        Mockito.when(client.getChannelManager()).thenReturn(manager);

        KaaClientState state = Mockito.mock(KaaClientState.class);
        KaaDataMultiplexer multiplexer = Mockito.mock(KaaDataMultiplexer.class);
        KaaDataDemultiplexer demultiplexer = Mockito.mock(KaaDataDemultiplexer.class);
        DefaultBootstrapChannelFake channel = new DefaultBootstrapChannelFake(client, state, failoverManager, 0);
        channel.setDemultiplexer(demultiplexer);
        channel.setMultiplexer(multiplexer);
        channel.shutdown();

        TransportConnectionInfo server = IPTransportInfoTest.createTestServerInfo(ServerType.BOOTSTRAP, TransportProtocolIdConstants.HTTP_TRANSPORT_ID,
                "localhost", 9889, KeyUtil.generateKeyPair().getPublic());
        channel.setServer(server);

        channel.sync(TransportType.BOOTSTRAP);
        channel.syncAll();

        channel.verify();
    }

}
